core/arch/arm/pta/system.c

```diff
@@ -168,11 +186,322 @@ static TEE_Result system_unmap(struct tee_ta_session *s, uint32_t param_types,
 			     ROUNDUP(params[0].value.a, SMALL_PAGE_SIZE));
 }
 
+static void ta_bin_close(void *ptr)
+{
+	struct bin_handle *binh = ptr;
+
+	if (binh) {
+		if (binh->op && binh->h)
+			binh->op->close(binh->h);
+		file_put(binh->f);
+	}
+	free(binh);
+}
+
+static TEE_Result system_open_ta_binary(struct system_ctx *ctx,
+					uint32_t param_types,
+					TEE_Param params[TEE_NUM_PARAMS])
+{
```

Comment:
I see your point and agree that's it's a problem, but I think it should be handled at a higher level or I fear that it will become a mess. Which libraries is the TA allowed to read? What about dlopen()?

I think we need a lightweight domain concept where a TA is allowed to read all TAs and libraries within. I'd prefer if I don't have to solve it in this PR, but I can try to add something crude if you think it's important.

Reply:
OK, I think this can wait, because:
1. The GlobalPlatform TEE Protection Profile (v1.2.1) does not seem to mandate confidentiality for TA code (only integrity). In other words, "TA separation" does not seem to encompass this aspect.
2. The vast majority of OP-TEE users must be using REE FS TAs which are not encrypted hence already exposed even to non-secure world.
3. Even if we assume that some users may have developed their own encrypted TA storage, the confidentiality of TA code remains with respect to normal world which is probably the main objective for implementing TA encryption.